numpy/_array_api/__init__.py

```diff
@@ -0,0 +1,165 @@
+  https://github.com/data-apis/array-api-tests. The test suite is still a work
+  in progress, but the existing tests pass on this module, with a few
+  exceptions:
+
+  - Device support is not yet implemented in NumPy
+    (https://data-apis.github.io/array-api/latest/design_topics/device_support.html).
+    As a result, the `device` attribute of the array object is missing, and
+    array creation functions that take the `device` keyword argument will fail
+    with NotImplementedError.
+
+  - DLPack support (see https://github.com/data-apis/array-api/pull/106) is
+    not included here, as it requires a full implementation in NumPy proper
+    first.
+
+  - np.argmin and np.argmax do not implement the keepdims keyword argument.
+
+  - Some linear algebra functions in the spec are still a work in progress (to
+    be added soon). These will be updated once the spec is.
+
+  - Some tests in the test suite are still not fully correct in that they test
+    all datatypes whereas certain functions are only defined for a subset of
+    datatypes.
+
+  The test suite is yet complete, and even the tests that exist are not
+  guaranteed to give a comprehensive coverage of the spec. Therefore, those
+  reviewing this submodule should refer to the standard documents themselves.
+
+- There is a custom array object, numpy._array_api.ndarray, which is returned
+  by all functions in this module. All functions in the array API namespace
+  implicitly assume that they will only receive this object as input. The only
+  way to create instances of this object is to use one of the array creation
+  functions. It does not have a public constructor on the object itself. The
+  object is a small wrapper Python class around numpy.ndarray. The main
+  purpose of it is to restrict the namespace of the array object to only those
+  dtypes and only those methods that are required by the spec, as well as to
+  limit/change certain behavior that differs in the spec. In particular:
+
+  - Indexing: Only a subset of indices supported by NumPy are required by the
+    spec. The ndarray object restricts indexing to only allow those types of
+    indices that are required by the spec. See the docstring of the
+    numpy._array_api.ndarray._validate_indices helper function for more
+    information.
+
+  - Type promotion: Some type promotion rules are different in the spec. In
+    particular, the spec does not have any value-based casting. Note that the
+    code to correct the type promotion rules on numpy._array_api.ndarray is
+    not yet implemented.
+
+- All functions include type annotations, corresponding to those given in the
+  spec (see _types.py for definitions of the types 'array', 'device', and
+  'dtype'). These do not currently fully pass mypy due to some limitations in
+  mypy.
+
+- The wrapper functions in this module do not do any type checking for things
+  that would be impossible without leaving the _array_api namespace.
```

Comment:
Could you give one such example here?

Reply:
It's actually not as bad as it sounds. Because we have a custom array class, most functions use `x._array` to access the underlying NumPy array, so if you pass in any other object, you will immediately get an AttributeError. The one example I can think of is that since we just reuse the NumPy dtype objects, you can slip a non-spec dtype through in some cases. It wouldn't work for any function that only accepts a subset of dtypes (like all the elementwise functions), since they check the input dtype against a list of allowed types. But for example this works:

```py
>>> import numpy as np
>>> import numpy._array_api as npa
>>> npa.ones((1,), dtype=np.float128)
ndarray([1.], dtype=float128)
```

The resulting object won't actually be guaranteed to work with any array API functions (and indeed, it won't work with many because it won't recognize that `np.float128` is actually a floating-point dtype).

A user is also not prevented from manually creating an invalid `np._array_api.ndarray` object by manually calling the private `_new` method. But the point here is that this sort of thing *is* impossible if you only use public methods in the array API namespace itself.